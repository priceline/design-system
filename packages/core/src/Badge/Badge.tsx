import themeGet from '@styled-system/theme-get'
import React from 'react'
import styled, { css } from 'styled-components'
import { SpaceProps, borderRadius, compose, space } from 'styled-system'
import type { ColorSchemeName } from '../theme/theme'
import { borderRadiusAttrs } from '../utils/attrs/borderRadiusAttrs'
import { applySizes, applyVariations, color, colorScheme, textTransform } from '../utils/utils'

const type = (props) => {
  const badgeColors = {
    blue: {
      backgroundColor: props.theme.colors.blue,
      color: props.theme.colors.white,
    },
    lightBlue: {
      backgroundColor: props.theme.colors.lightBlue,
      color: props.theme.colors.darkBlue,
    },
    green: {
      backgroundColor: props.theme.colors.green,
      color: props.theme.colors.white,
    },
    lightGreen: {
      backgroundColor: props.theme.colors.lightGreen,
      color: props.theme.colors.darkGreen,
    },
    red: {
      backgroundColor: props.theme.colors.red,
      color: props.theme.colors.white,
    },
    lightRed: {
      backgroundColor: props.theme.colors.lightRed,
      color: props.theme.colors.darkRed,
    },
    orange: {
      backgroundColor: props.theme.colors.orange,
      color: props.theme.colors.text,
    },
    gray: {
      backgroundColor: props.theme.colors.gray,
      color: props.theme.colors.white,
    },
    lightGray: {
      backgroundColor: props.theme.colors.lightGray,
      color: props.theme.colors.text,
    },
  }
  return !(props.bg && props.color) && (badgeColors[props.bg] || badgeColors.lightGray)
}

export const sizes = {
  captionSmall: css`
    ${themeGet('typography.caption')}
    padding: 2px 2px;
  `,
  captionMedium: css`
    ${themeGet('typography.caption')}
    padding: ${(props) => themeGet('space')(props)[1]};
  `,
  small: css`
    ${themeGet('typography.label')}
    padding: 3px 3px;
  `,
  medium: css`
    ${themeGet('typography.captionBold')}
    padding: ${(props) => themeGet('space')(props)[1]};
  `,
}

const letterSpacing = (props) => {
  return props.textTransform && props.textTransform !== 'uppercase'
    ? { letterSpacing: themeGet('letterSpacings.normal')(props) }
    : { letterSpacing: themeGet('letterSpacings.caps')(props) }
}

/**
 * @public
 */
export type BadgeProps = SpaceProps &
  React.HtmlHTMLAttributes<HTMLElement> & {
    size?: 'captionSmall' | 'captionMedium' | 'small' | 'medium'
    color?: string
    bg?: string
    borderRadius?: string
    colorScheme?: ColorSchemeName
    textTransform?: string
  }

/**
 * @public
 */
export const Badge: React.FC<BadgeProps> = styled.div.attrs(borderRadiusAttrs)`
  display: inline-block;
  ${({ theme }) => applySizes(sizes, undefined, theme.mediaQueries)};
  ${applyVariations('Badge')};
  ${type}
  ${color}
  ${colorScheme}
  ${textTransform}
  ${letterSpacing}

  ${(props) => compose(space, borderRadius)(props)}
`

Badge.defaultProps = {
  size: 'medium',
  px: 2,
  borderRadius: 'full',
  textTransform: 'uppercase',
}

Badge.displayName = 'Badge'
