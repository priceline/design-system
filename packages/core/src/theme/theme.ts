/* eslint-disable @typescript-eslint/naming-convention */

import { createTextStyles } from '../utils/createTextStyles'
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  FlexboxProps,
  GridProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps,
  space as _space,
  typography as _typography,
  background,
  border,
  color,
  compose,
  flexbox,
  grid,
  layout,
  position,
  shadow,
} from 'styled-system'
import { createColorStyles } from '../utils/createColorStyles'
import { createMediaQueries } from '../utils/createMediaQueries'
import { colors } from './colors'

const addAliases = (arr, aliases) =>
  aliases.forEach((key, i) =>
    Object.defineProperty(arr, key, {
      enumerable: false,
      get() {
        return this[i]
      },
    })
  )

/**
 * @public
 */
export { colors }

/**
 * @public
 */
export const breakpoints = [32, 40, 48, 64, 80].map((n) => n + 'em')

/**
 * @public
 */
export const mediaQueries = createMediaQueries(breakpoints)

const aliases = ['sm', 'md', 'lg', 'xl', 'xxl']

addAliases(breakpoints, aliases)
addAliases(mediaQueries, aliases)

/**
 * @public
 */
export const spaceValues = [0, 4, 8, 16, 32, 64, 128]
export const space = spaceValues.map((n) => n + 'px')

/**
 * @public
 */
export const font = `'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif`

/**
 * @public
 */
export const fontSizes = [12, 14, 16, 20, 24, 32, 40, 56, 72]

/**
 * @public
 */
export const medium = 500

/**
 * @public
 */
export const bold = 700

/**
 * @public
 */
export const regular = medium

/**
 * styled-system's `fontWeight` function can hook into the `fontWeights` object
 *
 * @public
 */
export const fontWeights = {
  medium,
  bold,
  // alias
  regular,
}

/**
 * @public
 */
export const lineHeights = {
  standard: 1.4,
  display: 1.25,
}

const letterSpacings = {
  normal: 'normal',
  caps: '0.025em',
}

/**
 * @public
 */
export const typography = {
  title1: {
    fontWeight: 700,
    fontSize: '80px',
    lineHeight: '96px',
  },
  title2: {
    fontWeight: 700,
    fontSize: '64px',
    lineHeight: '80px',
  },
  title3: {
    fontWeight: 700,
    fontSize: '52px',
    lineHeight: '64px',
  },
  heading1: {
    fontWeight: 700,
    fontSize: '40px',
    lineHeight: '48px',
  },
  heading2: {
    fontWeight: 700,
    fontSize: '32px',
    lineHeight: '40px',
  },
  heading3: {
    fontWeight: 700,
    fontSize: '26px',
    lineHeight: '32px',
  },
  heading4: {
    fontWeight: 700,
    fontSize: '20px',
    lineHeight: '24px',
  },
  heading5: {
    fontWeight: 700,
    fontSize: '16px',
    lineHeight: '20px',
  },
  heading6: {
    fontWeight: 700,
    fontSize: '14px',
    lineHeight: '16px',
  },
  article: {
    fontWeight: 500,
    fontSize: '16px',
    lineHeight: '28px',
    letterSpacing: '-0.006rem',
  },
  paragraph: {
    fontWeight: 500,
    fontSize: '14px',
    lineHeight: '20px',
    letterSpacing: '-0.003rem',
  },
  caption: {
    fontWeight: 500,
    fontSize: '12px',
    lineHeight: '16px',
    letterSpacing: '0.006rem',
  },
  overline: {
    fontWeight: 700,
    fontSize: '12px',
    lineHeight: '16px',
    letterSpacing: '0.063rem',
  },
  disclaimer: {
    fontWeight: 500,
    fontSize: '11px',
    lineHeight: '16px',
    letterSpacing: '0.013rem',
  },
  highlight: {
    fontWeight: 500,
    fontSize: '10px',
    lineHeight: '12px',
    letterSpacing: '0.013rem',
    textTransform: 'uppercase',
    caps: true,
  },
  highlightBold: {
    fontWeight: 700,
    fontSize: '10px',
    lineHeight: '12px',
    letterSpacing: '0.013rem',
    textTransform: 'uppercase',
    caps: true,
  },
  subtitle1: {
    fontWeight: 500,
    fontSize: '52px',
    lineHeight: '64px',
  },
  subheading1: {
    fontWeight: 500,
    fontSize: '40px',
    lineHeight: '48px',
  },
  subheading2: {
    fontWeight: 500,
    fontSize: '32px',
    lineHeight: '40px',
  },
  subheading3: {
    fontWeight: 500,
    fontSize: '26px',
    lineHeight: '32px',
  },
  subheading4: {
    fontWeight: 500,
    fontSize: '20px',
    lineHeight: '24px',
  },
  subheading5: {
    fontWeight: 500,
    fontSize: '16px',
    lineHeight: '20px',
  },
  subheading6: {
    fontWeight: 500,
    fontSize: '14px',
    lineHeight: '16px',
  },
  articleBold: {
    fontWeight: 700,
    fontSize: '16px',
    lineHeight: '28px',
    letterSpacing: '-0.006rem',
  },
  paragraphBold: {
    fontWeight: 700,
    fontSize: '14px',
    lineHeight: '20px',
    letterSpacing: '-0.003rem',
  },
  paragraph2: {
    fontWeight: 500,
    fontSize: '16px',
    lineHeight: '24px',
    letterSpacing: '-0.004rem',
  },
  paragraph2Bold: {
    fontWeight: 700,
    fontSize: '16px',
    lineHeight: '24px',
    letterSpacing: '-0.004rem',
  },
  captionBold: {
    fontWeight: 700,
    fontSize: '12px',
    lineHeight: '16px',
    letterSpacing: '0.006rem',
  },
  overlineMedium: {
    fontWeight: 500,
    fontSize: '12px',
    lineHeight: '16px',
    letterSpacing: '0.063rem',
  },
  disclaimerBold: {
    fontWeight: 700,
    fontSize: '11px',
    lineHeight: '16px',
    letterSpacing: '0.013rem',
  },
  label: {
    fontWeight: 700,
    fontSize: '10px',
    lineHeight: 1.4,
    letterSpacing: '0.2px',
  },
}

/**
 * @public
 */
export type ZIndexStrings =
  | 'auto'
  | 'absolute'
  | 'dropdown'
  | 'sticky'
  | 'fixed'
  | 'overlay'
  | 'onOverlay'
  | 'offCanvas'
  | 'modal'
  | 'popover'
  | 'tooltip'

/**
 * @public
 */
export type ZIndex = number | ZIndexStrings

/**
 * @public
 */
export type ZIndices = Record<ZIndexStrings, string | number>

/**
 * @public
 */
export const zIndices: ZIndices = {
  auto: 'auto',
  absolute: 50,
  dropdown: 100,
  sticky: 200,
  fixed: 300,
  tooltip: 400,
  offCanvas: 500,
  overlay: 550,
  onOverlay: 600,
  popover: 700,
  modal: 800,
}

/**
 * styled-system's `borderRadius` function can hook into the `radii` object/array
 * @public
 */
export const radii = [0, 2, 6]

/**
 * @public
 */
export const radius = '2px'

/**
 * @public
 */
export const baseBorderRadii = {
  none: 'none',
  xsm: '2px',
  sm: '4px',
  md: '8px',
  lg: '12px',
  xl: '16px',
  '2xl': '24px',
  '3xl': '32px',
  full: '9999px',
}

/**
 * @public
 */
export const actionBorderRadii = {
  'action-sm': '9999px',
  'action-md': '9999px',
  'action-lg': '12px',
  'action-xl': '12px',
}

/**
 * @public
 */
export const borderRadii = {
  ...baseBorderRadii,
  ...actionBorderRadii,
}

/**
 * @public
 */
export type BorderRadius = keyof typeof borderRadii

/**
 * @public
 */
export const maxContainerWidth = '1280px'

/**
 * @public
 */
export const shadows = {
  sm: '0 -1px 0 0 rgba(0,0,0,0.03),0 0 1px 0 rgba(0,0,0,0.24),0 2px 1px -1px rgba(0,0,0,0.16),0 2px 4px 0 rgba(0,0,0,0.12)',
  md: '0 -1px 0 0 rgba(0,0,0,0.03),0 0 2px 0 rgba(0,0,0,0.2),0 4px 2px -2px rgba(0,0,0,0.12),0 4px 8px -1px rgba(0,0,0,0.16)',
  lg: '0 -1px 0 0 rgba(0,0,0,0.03),0 1px 4px 0 rgba(0,0,0,0.2),0 6px 4px -4px rgba(0,0,0,0.12),0 8px 16px -1px rgba(0,0,0,0.16)',
  xl: '0 -1px 0 0 rgba(0,0,0,0.03),0 2px 8px 0 rgba(0,0,0,0.16),0 10px 8px -5px rgba(0,0,0,0.16),0 12px 32px -2px rgba(0,0,0,0.16)',
  '2xl':
    '0 -1px 0 0 rgba(0,0,0,0.03),0 4px 12px 0 rgba(0,0,0,0.16),0 12px 12px -4px rgba(0,0,0,0.16),0 24px 64px -2px rgba(0,0,0,0.16)',
  'overlay-md':
    '0 -1px 0 0 rgba(0,0,0,0.03),0 4px 16px 0 rgba(0,0,0,0.2),0 4px 16px -4px rgba(0,0,0,0.2),0 12px 24px -2px rgba(0,0,0,0.2)',
  'overlay-lg':
    '0 -1px 0 0 rgba(0,0,0,0.03),0 8px 32px 0 rgba(0,0,0,0.24),0 8px 16px 0 rgba(0,0,0,0.2),0 24px 64px 0 rgba(0,0,0,0.2)',
  'overlay-xl':
    '0 -1px 0 0 rgba(0,0,0,0.03),0 24px 72px 0 rgba(0,0,0,0.48),0 8px 16px 0 rgba(0,0,0,0.12),0 24px 64px 0 rgba(0,0,0,0.2)',
}

/**
 * @public
 */
export type BoxShadowSize = keyof typeof shadows

/**
 * @public
 */
export const textShadows = {
  sm: `0 1px 2px rgba(0,0,0,0.5)`,
  md: `0 2px 4px rgba(0,0,0,0.5)`,
}

const opacity = {
  'opacity-0': 0.0,
  'opacity-10': 0.1,
  'opacity-20': 0.2,
  'opacity-30': 0.3,
  'opacity-40': 0.4,
  'opacity-50': 0.5,
  'opacity-60': 0.6,
  'opacity-70': 0.7,
  'opacity-80': 0.8,
  'opacity-90': 0.9,
  'opacity-100': 1.0,
}

const scale = {
  'scale-0': 0.0,
  'scale-2': 0.02,
  'scale-6': 0.06,
  'scale-8': 0.08,
  'scale-10': 0.1,
  'scale-20': 0.2,
  'scale-30': 0.3,
  'scale-40': 0.4,
  'scale-50': 0.5,
  'scale-60': 0.6,
  'scale-70': 0.7,
  'scale-80': 0.8,
  'scale-90': 0.9,
  'scale-100': 1.0,
}

/**
 * @public
 */
export const duration = {
  fast: `150ms`,
  normal: `300ms`,
  slow: `450ms`,
  slowest: `600ms`,
  'duration-100': `100ms`,
  'duration-200': `200ms`,
  'duration-250': `250ms`,
  'duration-300': `300ms`,
  'duration-450': `450ms`,
}

// animation easing curves
const easeInOut = 'cubic-bezier(0.5, 0, 0.25, 1)'
const easeOut = 'cubic-bezier(0, 0, 0.25, 1)'
const easeIn = 'cubic-bezier(0.5, 0, 1, 1)'
const standardProductive = 'cubic-bezier(0.2, 0, 0.38, 0.9)'
const entranceProductive = 'cubic-bezier(0, 0, 0.38, 0.9)'
const exitProductive = 'cubic-bezier(0.2, 0, 1, 0.9)'
const standardExpressive = 'cubic-bezier(0.4, 0.14, 0.3, 1)'
const entranceExpressive = 'cubic-bezier(0, 0, 0.3, 1)'
const exitExpressive = 'cubic-bezier(0.4, 0.14, 1, 1)'

/**
 * @public
 */
export const timingFunctions = {
  easeInOut,
  easeOut,
  easeIn,
  'standard-productive': standardProductive,
  'entrance-productive': entranceProductive,
  'exit-productive': exitProductive,
  'standard-expressive': standardExpressive,
  'entrance-expressive': entranceExpressive,
  'exit-expressive': exitExpressive,
}

// animation delay
const transitionDelays = {
  small: `60ms`,
  medium: `160ms`,
  large: `260ms`,
  xLarge: `360ms`,
}

/**
 * @public
 */
export const textStyles = createTextStyles({
  fontSizes,
  fontWeights,
  lineHeights,
  letterSpacings,
})

/**
 * @public
 */
export const colorStyles = createColorStyles({ colors })

/**
 * @public
 */
export const theme = {
  breakpoints,
  mediaQueries,
  space,
  font,
  fontSizes,
  fontWeights,
  lineHeights,
  letterSpacings,
  regular,
  bold,
  textStyles,
  typography,
  colors,
  colorStyles,
  borderRadii,
  radii,
  radius,
  textShadows,
  shadows,
  maxContainerWidth,
  opacity,
  scale,
  duration,
  timingFunctions,
  transitionDelays,
  zIndices,
}

/**
 * @public
 */
export type TextStyle = {
  fontWeight: number
  fontSize: string
  lineHeight: string
}

/**
 * @public
 */
export const paletteFamilyVariations = [
  'lightest',
  'light',
  'tint',
  'base',
  'heading',
  'tone',
  'dark',
  'shade',
  'darkest',
] as const

/**
 * @public
 */
export type PaletteFamily = {
  lightest?: string
  light?: string
  tint?: string
  base: string
  heading?: string
  tone?: string
  dark?: string
  shade?: string
  darkest?: string
}

/**
 * @public
 */
export const paletteFamilyNames = [
  'primary',
  'secondary',
  'text',
  'highlight',
  'success',
  'error',
  'warning',
  'alert',
  'caution',
  'notify',
  'pricePrimary',
  'priceSecondary',
  'strike',
  'promoPrimary',
  'promoSecondary',
  'border',
  'background',
] as const

/**
 * @public
 */
export type PaletteFamilyVariation = (typeof paletteFamilyVariations)[number]

/**
 * @public
 */
export type PaletteFamilyName = (typeof paletteFamilyNames)[number]

/**
 * @public
 */
export type PaletteFamilies = Record<PaletteFamilyName, PaletteFamily>

/**
 * @public
 */
export type PaletteColor = `${PaletteFamilyName}.${PaletteFamilyVariation}`

export const paletteColors = paletteFamilyNames.flatMap((family) =>
  paletteFamilyVariations.map((variation) => `${family}.${variation}`)
)

/**
 * @public
 */
export type ColorStyle = {
  color: string
  backgroundColor: string
}

/**
 * @public
 */
export const colorStyleNames = [
  'whiteOnText',
  'whiteOnGray',
  'textOnLightGray',
  'whiteOnBlue',
  'blueOnLightBlue',
  'whiteOnGreen',
  'greenOnLightGreen',
  'whiteOnRed',
  'redOnLightRed',
  'textOnOrange',
  'whiteOnPurple',
  'purpleOnLightPurple',
  'textOnWhite',
  'grayOnWhite',
  'blueOnWhite',
  'greenOnWhite',
  'redOnWhite',
  'purpleOnWhite',
  'whiteOnDarkOrange',
  'info',
  'success',
  'warning',
  'danger',
] as const

/**
 * @public
 */
export type ColorStyleName = (typeof colorStyleNames)[number]

/**
 * @public
 */
export type ColorStyles = Record<ColorStyleName, ColorStyle>

/**
 * @public
 */
export type ColorScheme = {
  foreground: string
  background: string
  backgroundName: PaletteColor
}

/**
 * @public
 */
export const colorSchemeNames = [
  'primary',
  'primaryLight',
  'primaryLightest',
  'primaryDark',
  'primaryShade',
  'primaryDarkOnLight',
  'secondary',
  'secondaryLight',
  'secondaryLightest',
  'secondaryDark',
  'secondaryDarkOnLight',
  'neutral',
  'neutralLight',
  'neutralLightest',
  'neutralDark',
  'neutralDarkOnLight',
  'neutralDarkOnLightest',
  'success',
  'successLight',
  'successLightest',
  'successDark',
  'successDarkOnLight',
  'warning',
  'warningLight',
  'warningLightest',
  'cautionLight',
  'highlightLight',
  'promo',
  'promoLight',
  'promoLightest',
  'promoDark',
  'promoDarkOnLight',
  'alert',
  'alertLight',
  'alertLightest',
  'alertDarkOnLight',
  'heading',
  'headingBase',
  'headingLight',
  'headingOnPrimaryLight',
  'promoSecondary',
  'promoSecondaryDark',
] as const

/**
 * @public
 */
export type ColorSchemeName = (typeof colorSchemeNames)[number]

/**
 * @public
 */
export type ColorSchemes = Record<ColorSchemeName, ColorScheme>

/**
 * @public
 */
export type ColorName = keyof typeof colors

/**
 * @public
 */
export type ColorProp = ColorName | PaletteFamilyName | PaletteColor

/**
 * @public
 */
export type DesignSystemTheme = {
  space: string[]
  colors: Record<string, string>
  componentStyles: Record<string, Record<string, Record<string, Record<string, unknown>>>>
  mediaQueries: string[]
  breakpoints: string[]
  palette: PaletteFamilies
  colorStyles: ColorStyles
  textStyles: TextStyle[]
  contrastRatio: number
  colorSchemes: ColorSchemes
  zIndices: ZIndices
}

/**
 * @public
 */
export interface IStyledSystemProps
  extends BackgroundProps,
    BorderProps,
    ColorProps,
    FlexboxProps,
    GridProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {
  style?: React.CSSProperties
}

/**
 * @public
 */
export const ComposedStyleFns = compose(
  background,
  border,
  color,
  flexbox,
  grid,
  layout,
  position,
  shadow,
  _space,
  _typography
)
