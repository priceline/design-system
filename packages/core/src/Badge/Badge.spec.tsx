import React from 'react'
import { render, screen } from '../__test__/testing-library'
import { theme } from '../theme'
import { Badge } from './Badge'

describe('Badge', () => {
  let consoleError
  beforeEach(() => {
    consoleError = console.error
    console.error = jest.fn()
  })
  afterEach(() => (console.error = consoleError))

  test('renders', () => {
    const { asFragment } = render(<Badge />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders small', () => {
    const { asFragment } = render(<Badge size='small' />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('bg green sets background-color and color', () => {
    render(
      <Badge bg='green' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.green)
    expect(element).toHaveStyleRule('color', theme.colors.white)
  })

  test('bg lightGreen sets background-color and color', () => {
    render(
      <Badge bg='lightGreen' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.lightGreen)
    expect(element).toHaveStyleRule('color', theme.colors.darkGreen)
  })

  test('bg red sets background-color and color', () => {
    render(
      <Badge bg='red' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.red)
    expect(element).toHaveStyleRule('color', theme.colors.white)
  })

  test('bg lightRed sets background-color and color', () => {
    render(
      <Badge bg='lightRed' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.lightRed)
    expect(element).toHaveStyleRule('color', theme.colors.darkRed)
  })

  test('bg orange sets background-color and color', () => {
    render(
      <Badge bg='orange' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.orange)
    expect(element).toHaveStyleRule('color', theme.colors.text)
  })

  test('bg blue sets background-color and color', () => {
    render(
      <Badge bg='blue' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.blue)
    expect(element).toHaveStyleRule('color', theme.colors.white)
  })

  test('bg lightBlue sets background-color and color', () => {
    render(
      <Badge bg='lightBlue' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.lightBlue)
    expect(element).toHaveStyleRule('color', theme.colors.darkBlue)
  })

  test('non-preset: bg text sets background-color and color white sets color', () => {
    render(
      <Badge bg='text' color='white' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.text)
    expect(element).toHaveStyleRule('color', theme.colors.white)
  })

  test('can escape preset: bg lightBlue sets background-color and color text sets color', () => {
    render(
      <Badge bg='lightBlue' color='text' data-testid='badge'>
        Badge
      </Badge>
    )
    const element = screen.getByTestId('badge')
    expect(element).toHaveStyleRule('background-color', theme.colors.lightBlue)
    expect(element).toHaveStyleRule('color', theme.colors.text)
  })

  test('textTransform', () => {
    render(<Badge textTransform='lowercase'>Lowercase Text</Badge>)
    expect(screen.getByText('Lowercase Text')).toHaveStyleRule('text-transform', 'lowercase')
    expect(screen.getByText('Lowercase Text')).toHaveStyleRule('letter-spacing', 'normal')
  })

  test('size', () => {
    render(<Badge size='captionMedium'>Caption</Badge>)
    expect(screen.getByText('Caption')).toHaveStyleRule('padding-right', theme.space[2])
  })
})
