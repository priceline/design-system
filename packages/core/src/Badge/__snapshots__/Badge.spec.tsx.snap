// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge bg blue sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #0068ef;
  color: #fff;
  background-color: #0068ef;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge bg green sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #0a0;
  color: #fff;
  background-color: #0a0;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge bg lightBlue sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #e8f2ff;
  color: #049;
  background-color: #e8f2ff;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge bg lightGreen sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #ecf7ec;
  color: #060;
  background-color: #ecf7ec;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge bg lightRed sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #fbebeb;
  color: #800;
  background-color: #fbebeb;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge bg orange sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #f06f20;
  color: #001833;
  background-color: #f06f20;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge bg red sets background-color and color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #c00;
  color: #fff;
  background-color: #c00;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge can escape preset: bg lightBlue sets background-color and color text sets color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #e8f2ff;
  color: #001833;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  color="text"
  size="medium"
/>
`;

exports[`Badge non-preset: bg text sets background-color and color white sets color 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #001833;
  color: #fff;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  color="white"
  size="medium"
/>
`;

exports[`Badge renders 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 12px;
  line-height: 16px;
  -webkit-letter-spacing: 0.006rem;
  -moz-letter-spacing: 0.006rem;
  -ms-letter-spacing: 0.006rem;
  letter-spacing: 0.006rem;
  padding: 4px;
  background-color: #f4f6f8;
  color: #001833;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="medium"
/>
`;

exports[`Badge renders small 1`] = `
.c0 {
  display: inline-block;
  font-weight: 700;
  font-size: 10px;
  line-height: 1.4;
  -webkit-letter-spacing: 0.2px;
  -moz-letter-spacing: 0.2px;
  -ms-letter-spacing: 0.2px;
  letter-spacing: 0.2px;
  padding: 3px 3px;
  background-color: #f4f6f8;
  color: #001833;
  text-transform: uppercase;
  -webkit-letter-spacing: 0.025em;
  -moz-letter-spacing: 0.025em;
  -ms-letter-spacing: 0.025em;
  letter-spacing: 0.025em;
  padding-left: 8px;
  padding-right: 8px;
  border-radius: 9999px;
}

<div
  className="c0"
  size="small"
/>
`;
